plugins/widget/plugin.js

```diff
@@ -2053,9 +2053,62 @@
 	//
 	// @param {CKEDITOR.htmlParser.element} el
 	function cleanUpWidgetElement( el ) {
+		// Preserve initial and trailing space by replacing white space with &nbsp; (#605).
+		var firstTextNode = getFirstTextNode( el ),
+			lastTextNode = getLastTextNode( el );
+
+		if ( firstTextNode || lastTextNode && el.attributes[ 'data-cke-widget-data' ] ) {
+			// Add attribute to element if any white space will be replaced.
+			if ( firstTextNode.value.match( /^\s/g ) || lastTextNode.value.match( /\s$/g ) ) {
+				el.attributes[ 'data-cke-widget-white-space' ] = 1;
+			}
+
+			// Check whether the value of the first text node contains white space at the beginning and replace it with &nbsp;.
+			if ( firstTextNode.value && firstTextNode.value.match( /^\s/g ) ) {
+				firstTextNode.parent.attributes[ 'data-cke-white-space-first' ] = 1;
+				firstTextNode.value = firstTextNode.value.replace( /^\s/g, '&nbsp;' );
+			}
+
+			// Check whether the value of the last text node contains white space at the end and replace it with &nbsp;.
+			if ( lastTextNode.value && lastTextNode.value.match( /\s$/g ) ) {
+				lastTextNode.parent.attributes[ 'data-cke-white-space-last' ] = 1;
+				lastTextNode.value = lastTextNode.value.replace( /\s$/g, '&nbsp;' );
+			}
+		}
```

Comment:
This whole code block contains some redundant parts (checking for spaces in text node), I suggest to modify it as follows:

First `if` checks only if it operates on widget element:
`if ( el.attributes[ 'data-cke-widget-data' ] ) {`

`firstTextNode` and `lastTextNode` declaration is inside the above `if` (as this are used only here) with a third variable like `spacesReplaced`:
```
if ( el.attributes[ 'data-cke-widget-data' ] ) {
    var firstTextNode = getFirstTextNode( el ),
        lastTextNode = getLastTextNode( el ),
        spacesReplaced = false;
```

In each if where space is replaced you should set `spacesReplaced` to true:
```
if ( firstTextNode && firstTextNode.value && firstTextNode.value.match( /^\s/g ) ) { // Remember to check firstTextNode as previous if was modified.
    firstTextNode.parent.attributes[ 'data-cke-white-space-first' ] = 1;
    firstTextNode.value = firstTextNode.value.replace( /^\s/g, '&nbsp;' );
    spacesReplaced = true;
}
```

Finally, add `data-attribute` to widget element:
```
if ( spacesReplaced ) {
    el.attributes[ 'data-cke-widget-white-space' ] = 1;
}
```

This makes this code longer (3 lines more), but should be more readable and does not repeat any checks. WDYT?